### ISSUE TITLE
Outer bags

### ISSUE THREAD

**Issue by MassimilianoGrassiDataScience** [interpretml/interpret#206]
I have a few questions regarding the `outer_bags` parameter:

- Is each subsample created like a bootstrap subsample, i.e., via random sampling with replacement so that each bagging sample has the same sample size?
- if I select `outer_bags=1`, is the original sample used in the training process?
- if the bagging sample creation works like in bootstrap resampling, would it be legit to set it at a large number (say 1000) and use it to calculate the n% CIs for each level/bin of each feature to determine if that bin contributes to a significant change from the average probability of the output (for example, in a binary classification problem)?

Thanks!

**Comment by interpret-ml** [comment]
Hi @MassimilianoGrassiDataScience, 

Good questions! The subsampling in `outer_bags` is done _without_ replacement. The algorithm randomly samples (1 - `validation_size`) datapoints as training data, and uses the rest as validation data per inner estimator (once per outer_bag). This is mostly done for computational efficiency, and our experiments show that it makes very little difference to the final models in practice. 

Setting `outer_bags=1` doesn't use the entire training sample -- the algorithm will still only use (1 - `validation_size`) of the data to train. An internal validation set is still needed for the early stopping mechanism in the model to work. However, setting `outer_bags=1` ensures that this process is only done once, and the single model trained will be equivalent to the final EBM. When `outer_bags>1`, the final model is an average of all models trained on different subsamples. 

Your third question is a bit trickier. First, setting the number really high is very computationally expensive -- each increment to `outer_bags` involves training an entirely new EBM. This process is parallelized across the number of cores on your machine, but 1000 runs might still be computationally infeasible on many datasets. In addition, calculating CIs directly on bins is difficult for a few reasons. Firstly, there is significant correlation between bins based on proximity due to EBM being a tree based algorithm. In addition, defining "significant change" is also difficult. The algorithm uses 256 bins by default on numeric features, which means simply testing for significance on each bin leads to a large [multiple hypothesis testing problem](https://en.wikipedia.org/wiki/Multiple_comparisons_problem). Even at strong significance thresholds, you're very likely to have bins marked as spuriously significant by chance, simply due to the number of tests running. The traditional ways of correcting for this (like [Bonferroni correction](https://en.wikipedia.org/wiki/Bonferroni_correction)) are simply far too aggressive in practice for our setting.

The uncertainty intervals shown on our visualizations in `explain_global()` are standard deviations of estimates per bin across each of the bags. While they are by no means "true" confidence intervals, they give some indication of how much model predictions can change with minor variations in the training data. We would not recommend using these as true CIs, and there is no statistical meaning attached to things like the interval crossing zero, but they can be used as rough heuristics. Calculating proper CIs for each bin is something we are very interested in, and we continue to research this further on our end. If you (or anyone reading this issue!) would like to collaborate on this, please let us know!

Thanks,
-InterpretML Team
